src/serac/physics/materials/hyperelastic_material.hpp

```diff
@@ -0,0 +1,293 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file hyperelastic_material.hpp
+ *
+ * @brief The hyperelastic material models for the solid module
+ */
+
+#pragma once
+
+#include "mfem.hpp"
+
+#include "serac/numerics/array_4D.hpp"
+
+namespace serac {
+
+/**
+ * @brief Abstract interface class for a generic hyperelastic material
+ *
+ */
+class HyperelasticMaterial {
+public:
+  /**
+   * @brief Construct a new Hyperelastic Material object
+   *
+   */
+  HyperelasticMaterial() : Ttr_(nullptr) {}
+
+  /**
+   * @brief Destroy the Hyperelastic Material object
+   *
+   */
+  virtual ~HyperelasticMaterial() = default;
+
+  /// A reference-element to target-element transformation that can be used to
+  /// evaluate mfem::Coefficient%s.
+  /** @note It is assumed that _Ttr.SetIntPoint() is already called for the
+      point of interest. */
+
+  /**
+   * @brief Set the reference-to-target transformation. This is required to use coefficient parameters.
+   *
+   * @param[in] Ttr The reference-to-target (stress-free) transformation
+   */
+  void SetTransformation(mfem::ElementTransformation& Ttr) { Ttr_ = &Ttr; }
+
+  /**
+   * @brief Evaluate the strain energy density function, W = W(F).
+   *
+   * @param[in] F The deformation gradient
+   * @return double Strain energy density
+   */
+  virtual double EvalStrainEnergy(const mfem::DenseMatrix& F) const = 0;
+
+  /**
+   * @brief Evaluate the Cauchy stress sigma = sigma(F).
```

Comment:
In some cases, say for anisotropic response, it might be more natural to define the stress in the reference configuration, i.e. to provide the PK1 stress rather than the Cauchy. Returning the PK1 stress would also make this function more consistent with the `AssembleTangentModuli` method. I understand there might be other cases where using the Cauchy stress is more natural. Could the `HyperelasticMaterial` class provide methods to map between these two stress tensors (and possibly others)?

Reply:
Related to the above and to the discussion [here](https://github.com/LLNL/serac/pull/353/files#r570538970), would it be possible to provide efficient and accurate "building blocks" like F, C, E, etc. that could be used to implement different material models?

Reply:
I'll add a kinematic helper function file to contain these transformations.

Reply:
It occurred to me that a cleaner way to address the stress definition might be to generalize the `compute_on_reference` `bool`s used in e.g. the `TractionIntegrator`. The `EvalStress` method could take two additional arguments, `bool area_on_reference = false, bool force_on_reference = false`. This would let the default return be the Cauchy stress, but allow an implementation to use the PK1 or PK2 stress if that makes more sense; depending on the value of the bools, the returned stress could be transformed into whatever stress measure the solver actually uses. This would also mean someone implementing something other than the Cauchy stress would not need to also implement the transforms themselves, removing a possible source of bugs.

Reply:
I worry about having a list of bools to define behavior in a function call as it is difficult to tell what `foo(bar, false, true)` does at a glance. An easier method is probably having some kind of annotation on the return type. Any thoughts @white238  @joshessman-llnl ?

Reply:
That's a fair point. Another possibility would be an enumeration of configurations, with default values for `EvalStress`. I think `foo(bar, ReferenceConfig, CurrentConfig)` would at least be easier to parse than the case with `bool`s. Annotating the return type would make it even easier to parse since you wouldn't have to remember the argument ordering. On the other hand, the `enum` could also be in the signature for defining BCs, which might improve their readability. Any solution works for me.

Reply:
If we introduce two bools/enum flags to the user interface for both the area and force, then this opens up the possibility for the user to request area in terms of the current config and force in terms of the reference config - is that something we'd want to allow?  

If not, maybe it'd be preferable to use an `enum StressType { Cauchy, PK1, PK2 };` in the interface and translate it into `area_on_ref`, `force_on_ref` internally if needed.

Reply:
After talking with @solberg2 and discussing this at the meeting, the consensus was we should keep the Cauchy stress interface for now as it is easier to debug issues at a material level than an assembly level. With @samuelpmishLLNL 's upcoming rewrite, adding options like this will be much more feasible.

Reply:
Realistically, any material that can be written in terms of P_K2, etc. could be calculated as such and then transformed "at the end".  You could even output the P_K2 stresses and strains as plot/history variables for debugging.  But best to communicate to the rest of the code in a consistent manner (e.g. the Cauchy Stress)